sqldelight-runtime/src/main/java/com/squareup/sqldelight/SqlDelightDatabase.kt

```
@@ -1,93 +1,76 @@
-      transactions.set(transaction.parent)
-      helper.writableDatabase.endTransaction()
-      if (transaction.commit) {
-        transaction.forEach { function -> afterTransaction(function) }
-      }
-    }
-
-    override fun close() {
-      end()
-    }
-  }
+  internal val helper: SqlDatabaseHelper
+) {
+  private val transactions = ThreadLocal<Transaction>()
+  private val actions = ArrayList<() -> Unit>()
 
-  override fun newTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListener(transaction)
-
-    return this.transaction
-  }
-
-  override fun newNonExclusiveTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListenerNonExclusive(transaction)
-
-    return this.transaction
-  }
-
-  override fun afterTransaction(function: () -> Unit) {
+  fun afterTransaction(function: () -> Unit) {
     val transaction = transactions.get()
     if (transaction != null) {
-      transaction.add(function)
+      actions.add(function)
     } else {
       function()
     }
   }
 
-  override fun close() {
+  fun close() {
     helper.close()
   }
-}
 
-private class SqliteTransaction(
-    val parent: SqliteTransaction?
-) : LinkedHashSet<() -> Unit>(), SQLiteTransactionListener {
-  var commit: Boolean = false
+  fun transaction(body: Transaction.() -> Unit) {
+    val parent = transactions.get()
+    val transaction = Transaction()
 
-  override fun onBegin() {}
+    if (parent == null) {
+      helper.getConnection().beginTransaction()
+    }
 
-  override fun onCommit() {
-    commit = true
-  }
+    try {
+      transactions.set(transaction)
+      transaction.body()
+      transactions.set(parent)
+    } catch(e: RollbackException) {
```

With `finally`, this should just become an empty catch